include/deal.II/base/lazy.h

```diff
@@ -314,72 +287,88 @@ inline Lazy<T>::Lazy()
 {}
 
 
+
 template <typename T>
 DEAL_II_CXX20_REQUIRES((std::is_move_constructible_v<T> &&
                         std::is_move_assignable_v<T>))
 inline Lazy<T>::Lazy(const Lazy &other)
-  : object(other.object)
 {
-  object_is_initialized.store(other.object_is_initialized.load());
+  // If the other object has a value stored, then get it and set our
+  // own object to a copy of it:
+  if (other.has_value())
+    {
+      object_is_initialized.store(true);
+      task_result =
+        Threads::new_task([&other]() -> T { return other.value(); });
+      task_result.join();
+    }
+  else
+    object_is_initialized.store(false);
 }
 
 
+
 template <typename T>
 DEAL_II_CXX20_REQUIRES((std::is_move_constructible_v<T> &&
                         std::is_move_assignable_v<T>))
 inline Lazy<T>::Lazy(Lazy &&other) noexcept
-  : object(std::move(other.object))
+  : task_result(std::move(other.task_result))
 {
   object_is_initialized.store(other.object_is_initialized.load());
 
-  // Mark the other object as uninitialized. This is marginally non-trivial
-  // because moving from std::optional<T> does *not* result in an empty
-  // std::optional<T> but instead one that does contain a T, but one that
-  // has been moved from -- typically something akin to a default-initialized
-  // T. That seems undesirable, so reset everything to an empty state.
+  // Mark the other object as uninitialized.
   other.object_is_initialized.store(false);
-  other.object.reset();
 }
 
 
+
 template <typename T>
 DEAL_II_CXX20_REQUIRES((std::is_move_constructible_v<T> &&
                         std::is_move_assignable_v<T>))
 inline Lazy<T> &Lazy<T>::operator=(const Lazy &other)
 {
-  object = other.object;
-  object_is_initialized.store(other.object_is_initialized.load());
+  // If the other object has a value stored, then get it and set our
+  // own object to a copy of it:
+  if (other.has_value())
+    {
+      object_is_initialized.store(true);
+      task_result =
```

Comment:
again, try_emplace_task()?

Reply:
Debatable here. First, I really do want to emplace a task because `operator=` should overwrite whatever we used to store. Second, `try_emplace_back()` is meant for situations where multiple threads at the same time try to access the `TaskResult` object. But you can't sensibly define the semantics of `Lazy::operator=` if multiple threads are calling this operator at the same time. That's *always* going to be a race condition. 

The way I see it is that we only need to worry about the correctness of multiple `const` functions running concurrently. I think that's a sensible guide to what we want to support. So `operator=`, not being `const`, is not something we need to ensure can be called multiple times concurrently.